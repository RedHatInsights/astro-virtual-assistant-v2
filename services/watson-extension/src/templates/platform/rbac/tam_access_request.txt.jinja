{%- if ok -%}
Your request for access to customer account {{account_id}} has been submitted. The customer will be notified shortly.
{%- else -%}
There was an error processing your request. Please try again later.
{%- endif -%}

Future requests can also be made using the form [here](/internal/access-requests).
