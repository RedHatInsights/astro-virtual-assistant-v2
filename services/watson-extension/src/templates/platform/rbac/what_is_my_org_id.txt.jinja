Your Org ID is {{org_id}}. You can also find it by clicking on your name at the top right of the screen.
