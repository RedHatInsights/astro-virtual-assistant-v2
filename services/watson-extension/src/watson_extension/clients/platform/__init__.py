from dataclasses import dataclass


@dataclass
class IntegrationInfo:
    name: str
    enabled: bool
    type: str
    group: str
    id: str
