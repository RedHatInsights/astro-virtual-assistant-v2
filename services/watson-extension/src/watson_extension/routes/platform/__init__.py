from quart import Blueprint
from . import chrome

blueprint = Blueprint("platform", __name__, url_prefix="/platform")

blueprint.register_blueprint(chrome.blueprint)
