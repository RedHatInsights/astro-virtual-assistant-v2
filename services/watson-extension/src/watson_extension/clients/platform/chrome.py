import abc
import json
import injector
from dataclasses import dataclass
from typing import List, Optional

from common.identity import AbstractUserIdentityProvider
from common.platform_request import AbstractPlatformRequest
from watson_extension.clients import ChromeServiceURL


@dataclass
class Favorite:
    id: str
    pathname: str
    favorite: bool
    user_identity_id: str


@dataclass
class User:
    account_id: str
    first_login: bool
    day_one: bool
    last_login: str
    last_visited_pages: List[dict]
    favorite_pages: List[Favorite]
    visited_bundles: dict


@dataclass
class Link:
    id: Optional[str]
    title: str
    alt_title: Optional[List[str]]
    links: Optional["Link"]
    app_id: Optional[str] = None
    is_group: bool = False
    is_external: bool = False
    href: Optional[str] = None


@dataclass
class Service:
    description: str
    id: str
    links: Optional[List[dict]]
    title: str
    href: Optional[str]


class ChromeServiceClient(abc.ABC):
    @abc.abstractmethod
    async def get_user(self) -> User: ...

    @abc.abstractmethod
    async def get_generated_services(self) -> List[Service]: ...

    @abc.abstractmethod
    async def modify_favorite_service(
        self, service, favorite=True
    ) -> List[Favorite]: ...


class ChromeServiceClientHttp(ChromeServiceClient):
    def __init__(
        self,
        chrome_url: injector.Inject[ChromeServiceURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.chrome_url = chrome_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def get_user(self):
        # struggles to be a json response, manually doing it here
        response = await self.platform_request.get(
            self.chrome_url,
            "/api/chrome-service/v1/user",
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        data = json.loads(await response.text())["data"]
        return User(
            account_id=data["accountId"],
            first_login=data["firstLogin"],
            day_one=data["dayOne"],
            last_login=data["lastLogin"],
            last_visited_pages=data["lastVisitedPages"],
            favorite_pages=[
                Favorite(
                    id=fp["id"],
                    pathname=fp["pathname"],
                    favorite=fp["favorite"],
                    user_identity_id=fp["userIdentityId"],
                )
                for fp in data.get("favoritePages", [])
            ],
            visited_bundles=data.get("visitedBundles", {}),
        )

    async def get_generated_services(self) -> List[Service]:
        response = await self.platform_request.get(
            self.chrome_url,
            "/api/chrome-service/v1/static/stable/prod/services/services-generated.json",
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        data = await response.json()
        return [
            Service(
                description=service.get("description", ""),
                id=service.get("id", ""),
                links=parse_links_into_obj(service.get("links", [])),
                title=service.get("title", ""),
                href=service.get("href", ""),
            )
            for service in data
        ]

    async def modify_favorite_service(self, href, favorite=True):
        response = await self.platform_request.post(
            self.chrome_url,
            "/api/chrome-service/v1/favorite-pages",
            user_identity=await self.user_identity_provider.get_user_identity(),
            json={
                "favorite": favorite,
                "pathname": href,
            },
        )
        response.raise_for_status()
        return await response.text()


def parse_links_into_obj(links) -> List[Link]:
    if not links:
        return []
    parsed_links = []
    for link in links:
        parsed_links.append(
            Link(
                id=link.get("id"),
                title=link.get("title", ""),
                alt_title=link.get("altTitle", []),
                links=parse_links_into_obj(link.get("links", [])),
                app_id=link.get("appId"),
                is_group=link.get("isGroup", False),
                is_external=link.get("isExternal", False),
                href=link.get("href", ""),
            )
        )
    return parsed_links
