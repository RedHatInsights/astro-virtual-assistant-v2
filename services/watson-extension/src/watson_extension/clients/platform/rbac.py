import abc
import injector
import logging
from dataclasses import dataclass
from typing import Any, Dict, List

from common.identity import AbstractUserIdentityProvider
from common.platform_request import AbstractPlatformRequest
from watson_extension.clients import RbacURL
from watson_extension.config import is_running_locally


logger = logging.getLogger(__name__)


@dataclass
class Roles:
    uuid: str
    name: str
    display_name: str
    description: str
    created: str
    modified: str
    policyCount: int
    groups_in_count: int
    accessCount: int
    applications: List[str]
    system: bool
    platform_default: bool
    admin_default: bool
    external_role_id: str = None
    external_tenant: str = None


@dataclass
class TAMRequestAccessPayload:
    account_id: str
    org_id: str
    start_date: str
    end_date: str
    roles: List[str]


class RBACClient(abc.ABC):
    @abc.abstractmethod
    async def get_roles_for_tam(self) -> List[Roles]: ...

    async def send_rbac_tam_request(self, body: Dict[str, Any]): ...


class RBACClientHttp(RBACClient):
    def __init__(
        self,
        rbac_url: injector.Inject[RbacURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.rbac_url = rbac_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def get_roles_for_tam(self) -> List[Roles]:
        """Get roles for TAM"""
        response = await self.platform_request.get(
            self.rbac_url,
            "/api/rbac/v1/roles/?system=true&limit=9999&order_by=display_name&add_fields=groups_in_count",
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()
        return [
            Roles(
                uuid=role["uuid"],
                name=role["name"],
                display_name=role["display_name"],
                description=role["description"],
                created=role["created"],
                modified=role["modified"],
                policyCount=role["policyCount"],
                groups_in_count=role["groups_in_count"],
                accessCount=role["accessCount"],
                applications=role["applications"],
                system=role["system"],
                platform_default=role["platform_default"],
                admin_default=role["admin_default"],
                external_role_id=role.get("external_role_id"),
                external_tenant=role.get("external_tenant"),
            )
            for role in response.json()
        ]

    async def send_rbac_tam_request(self, body: TAMRequestAccessPayload):
        if is_running_locally:
            logger.info(
                f"Called send_rbac_tam_request in local environment with body: {body}"
            )

            from unittest.mock import Mock

            mock_response = Mock()
            mock_response.ok = True

            return mock_response

        # POST https://console.stage.redhat.com/api/rbac/v1/cross-account-requests/
        return await self.platform_request.post(
            self.rbac_url,
            "/api/rbac/v1/cross-account-requests/",
            user_identity=await self.user_identity_provider.get_user_identity(),
            json=body,
        )
