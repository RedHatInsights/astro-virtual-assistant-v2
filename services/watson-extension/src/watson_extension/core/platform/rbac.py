from typing import Any, Dict, List, Tuple
import injector

from watson_extension.clients.platform.rbac import (
    RBACClient,
    Roles,
    TAMRequestAccessPayload,
)

_DURATIONS = ["3 days", "1 week", "2 weeks"]


class RBACCore:
    def __init__(self, rbac_client: injector.Inject[RBACClient]):
        self.rbac_client = rbac_client

    async def get_roles_for_tam(self) -> List[Dict[str, Any]]:
        return await self.rbac_client.get_roles_for_tam()

    async def send_rbac_tam_request(
        self, account_id, org_id, start_date, end_date, roles
    ) -> bool:
        return await self.rbac_client.send_rbac_tam_request(
            TAMRequestAccessPayload(
                account_id=account_id,
                org_id=org_id,
                start_date=start_date,
                end_date=end_date,
                roles=_get_roles_names(roles),
            )
        )

    def get_start_end_date_from_duration(self, duration: str) -> Tuple[str, str]:
        from datetime import date, timedelta

        start_date = date.today()
        duration_map = {
            "3 days": timedelta(days=3),
            "1 week": timedelta(weeks=1),
            "2 weeks": timedelta(weeks=2),
        }
        end_date = start_date + duration_map.get(duration, timedelta(0))

        return start_date.strftime("%m/%d/%Y"), end_date.strftime("%m/%d/%Y")


def _get_roles_names(roles: List[Roles]):
    return [role.display_name for role in roles]
