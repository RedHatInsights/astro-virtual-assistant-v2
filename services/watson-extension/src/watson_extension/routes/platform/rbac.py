import injector
from pydantic import BaseModel

from quart import Blueprint, render_template
from quart_schema import validate_response, validate_querystring, document_headers

from watson_extension.core.platform.rbac import (
    RBACCore,
)
from watson_extension.routes import RHSessionIdHeader

blueprint = Blueprint("rbac", __name__, url_prefix="/rbac")


class TamAccessRequestQuery(BaseModel):
    account_id: str
    org_id: str
    duration: str


class TamAccessRequestResponse(BaseModel):
    response: str


@blueprint.post("/tam-access")
@validate_querystring(TamAccessRequestQuery)
@validate_response(TamAccessRequestResponse)
@document_headers(RHSessionIdHeader)
async def send_tam_access(
    query_args: TamAccessRequestQuery,
    rbac_core: injector.Inject[RBACCore],
) -> TamAccessRequestResponse:
    start_date, end_date = rbac_core.get_start_end_date_from_duration(
        query_args.duration
    )
    roles = await rbac_core.get_roles_for_tam()
    response = await rbac_core.send_rbac_tam_request(
        query_args.account_id, query_args.org_id, start_date, end_date, roles
    )

    # TODO check if they are internal, need a helper?
    return TamAccessRequestResponse(
        response=await render_template(
            "platform/rbac/tam_access_request.txt.jinja",
            ok=response.ok,
            account_id=query_args.account_id,
        )
    )
