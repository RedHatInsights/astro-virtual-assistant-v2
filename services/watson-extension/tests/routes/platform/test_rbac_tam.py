from datetime import date, timedelta
from unittest.mock import MagicMock

import injector
import pytest
from quart.typing import TestClientProtocol

from watson_extension.clients.platform.rbac import (
    RBACClient,
    Roles,
)
from watson_extension.core.platform.rbac import RBACCore

from ..common import app_with_blueprint
from watson_extension.routes.platform.rbac import blueprint
from ... import async_value, get_test_template


@pytest.fixture
async def rbac_client() -> MagicMock:
    rbac_client = MagicMock(RBACClient)
    rbac_client.get_roles_for_tam = MagicMock(
        return_value=async_value(
            [
                Roles(
                    uuid="1001",
                    name="Automation Analytics Viewer",
                    display_name="Automation Analytics viewer",
                    description="Perform read operations on Automation Analytics resources.",
                    created="2021-06-14T13:53:00.990946Z",
                    modified="2024-09-19T20:07:20.563863Z",
                    policyCount=1562,
                    groups_in_count=3,
                    accessCount=1,
                    applications=["automation-analytics"],
                    system=True,
                    platform_default=False,
                    admin_default=False,
                    external_role_id=None,
                    external_tenant=None,
                ),
                Roles(
                    uuid="1002",
                    name="Foo Viewer",
                    display_name="Foo bariest",
                    description="Perform read operations on kung foo resources.",
                    created="2021-06-14T13:53:00.990946Z",
                    modified="2024-09-19T20:07:20.563863Z",
                    policyCount=1562,
                    groups_in_count=3,
                    accessCount=1,
                    applications=["foo-bar"],
                    system=True,
                    platform_default=False,
                    admin_default=False,
                    external_role_id=None,
                    external_tenant=None,
                ),
            ]
        )
    )
    return rbac_client


@pytest.fixture
async def test_client(rbac_client) -> TestClientProtocol:
    def injector_binder(binder: injector.Binder):
        binder.bind(RBACClient, rbac_client)

    return app_with_blueprint(blueprint, injector_binder).test_client()


@pytest.mark.asyncio
async def test_tam_request_network_error(test_client, rbac_client: MagicMock):
    rbac_client.get_roles_for_tam.side_effect = Exception("Mocked network error")

    with pytest.raises(Exception) as excinfo:
        await rbac_client.get_roles_for_tam()
    assert "Mocked network error" in str(excinfo.value)


async def test_tam_request_access(test_client, rbac_client) -> None:
    response = await test_client.post(
        "/rbac/tam-access",
        query_string={"account_id": "12345", "org_id": "foo", "duration": "3 days"},
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == get_test_template(
        "platform/rbac/submit_tam_access_request.txt"
    )

async def test_tam_request_non_200(test_client, rbac_client: MagicMock):
    rbac_client.send_rbac_tam_request.return_value = MagicMock(
        ok=False,
        status=400,
        text=async_value("Bad Request"),
    )
    response = await test_client.post(
        "/rbac/tam-access",
        query_string={"account_id": "12345", "org_id": "foo", "duration": "3 days"},
    )

    assert response.status_code == 200 # ignores that
    data = await response.get_json()
    assert data["response"] == get_test_template(
        "platform/rbac/error_tam_access_request.txt"
    )

@pytest.mark.asyncio
async def test_tam_request_body(test_client, rbac_client: MagicMock):
    async def mock_send_rbac_tam_request(payload):
        assert payload.account_id == "12345"
        assert payload.org_id == "foo"
        assert payload.start_date == date.today().strftime("%m/%d/%Y")
        assert payload.end_date == (date.today() + timedelta(days=3)).strftime("%m/%d/%Y")
        assert payload.roles == ["Automation Analytics viewer", "Foo bariest"]
        return MagicMock(ok=True, status=200, text=async_value("Success"))

    rbac_client.send_rbac_tam_request.side_effect = mock_send_rbac_tam_request

    response = await test_client.post(
        "/rbac/tam-access",
        query_string={"account_id": "12345", "org_id": "foo", "duration": "3 days"},
    )

    assert response.status == "200 OK"

    data = await response.get_json()
    assert data["response"] == get_test_template(
        "platform/rbac/submit_tam_access_request.txt"
    )

@pytest.mark.parametrize(
    "duration, expected_end_date",
    [
        ("3 days", date.today() + timedelta(days=3)),
        ("1 week", date.today() + timedelta(weeks=1)),
        ("2 weeks", date.today() + timedelta(weeks=2)),
        ("invalid_duration", date.today()),
    ],
)
def test_get_start_end_date_from_duration(test_client, rbac_client: MagicMock, duration, expected_end_date):
    start_date, end_date = RBACCore(rbac_client=rbac_client).get_start_end_date_from_duration(duration)
    assert start_date == date.today().strftime("%m/%d/%Y")
    assert end_date == expected_end_date.strftime("%m/%d/%Y")
