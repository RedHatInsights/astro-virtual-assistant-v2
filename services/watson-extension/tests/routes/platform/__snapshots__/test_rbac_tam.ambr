# serializer version: 1
# name: test_tam_request_access
  '''
  Your request for access to customer account 12345 has been submitted. The customer will be notified shortly.
  
  Future requests can also be made using the form [here](/internal/access-requests).
  '''
# ---
# name: test_tam_request_body
  '''
  Your request for access to customer account 12345 has been submitted. The customer will be notified shortly.
  
  Future requests can also be made using the form [here](/internal/access-requests).
  '''
# ---
# name: test_tam_request_non_200
  '''
  There was an error processing your request. Please try again later.
  
  Future requests can also be made using the form [here](/internal/access-requests).
  '''
# ---
